### ISSUE TITLE
LSP server in CudaText: inserting of a snippet don't replace the entire word

### ISSUE THREAD

**Issue by Alexey-T** [clangd/clangd#1302]
CudaText editor ( https://cudatext.github.io/ ) with LSP Client plugin.
Config for clangd is shown here:
https://wiki.freepascal.org/CudaText_plugins#LSP_servers_for_C.2FC.2B.2B

'new.cpp' file with C++ syntax
```
#include <stdio.h>

void main(){

}
```
Put caret inside 'void': `v|oid`.
Call autocompletion. Choose 'volatile' item.
Result: `volatileoid main(){`

Ie, server doesn't report the correct replacement range, so plugin doesn't replace correct range.
@veksha

**Comment by Alexey-T** [comment]
Ubuntu 20.04
```
user@PC:~$ clangd --version
Ubuntu clangd version 12.0.0-3ubuntu1~20.04.5
```

**Comment by veksha** [comment]
FYI
Sublime Text + clangd:
![image](https://user-images.githubusercontent.com/275333/191423773-ff18fdd4-63c0-42f2-b5b2-5bc393df6bed.png)

**Comment by HighCommander4** [comment]
Not specific to an editor, vscode has the same behaviour. #387 is related.

**Comment by sam-mccall** [comment]
Right, the server has a choice of how much text to the right of the cursor to eat when accepting a completion.

My experience is that (non-clangd) editors are inconsistent about this, and so I try to avoid the situation when editing. My guess is eating no characters is most common, but curious what other editors do.

Our general approach has been "there's no answer that feels obviously correct to everyone in every situation, so conservatively consume nothing". #387 identified some cases where we thought eating up to brackets felt safe.

> server doesn't report the correct replacement range, so plugin doesn't replace correct range

Can you explain why you think replacing "oid" is correct here? Would it still hold if the original next was "vvoid", so that the post edit options would be "volatilevoid" or "volatile"?

**Comment by i-ky** [comment]
> My guess is eating no characters is most common, but curious what other editors do.

For the record, in [Kate](https://kate-editor.org/) and related editors (KWrite, KDevelop, ...) it is (somewhat) [configurable](https://docs.kde.org/stable5/en/kate/katepart/config-dialog.html#pref-edit):
> **Remove tail on complete**
> Remove the tail of a previous word when the completion item is chosen from a list.

**Comment by veksha** [comment]
Hi, @sam-mccall,
In the past CudaText had client-side "entire world replacing" way of dealing with this issue, and I don't remember any problems like you describe -> _"incorrectly eating characters (deleted user code)"_
1. Undo/Redo features are builtin in every editor in the world.
2. replacing range is stretches up to the end of current "word" ("word" in terms of C++ lexer). server must know what is "word delimiters". every user code has them, so replacement range will never be huge, in fact it will be the smallest possible. (so don't worry about unwanted loss of code.)

3. additionally server may check for "()" or "(" and include that in replacing range.

**Comment by sam-mccall** [comment]
I do understand what you're asking for, what's missing is a rationale for this being better in general.

What is a user hoping to achieve when they trigger completion at `v|oid`? Are we confident enough in that guess to do something potentially surprising?

"This worked in CudaText, and people seemed happy enough" is something, but vim and VSCode both work this way by default, so by that measure the existing behavior is pretty popular too.

**Comment by HighCommander4** [comment]
See also #2190 which is about a mechanism added to LSP in version 3.16 which allows clients to (let the user) opt into "replace the entire word" behaviour.

**Comment by HighCommander4** [comment]
I think we can consider this resolved via #2190.
